src/main/scala/com/databricks/spark/avro/SchemaConverters.scala

```diff
@@ -264,12 +273,29 @@ object SchemaConverters {
                   case f: java.lang.Float => new java.lang.Double(f.doubleValue())
                 }
               }
-            case other => throw new IncompatibleSchemaException(
-              s"Cannot convert Avro schema to catalyst type because schema at path " +
-                s"${path.mkString(".")} is not compatible (avroType = $other, sqlType = $sqlType)" +
-                s" or this mix of union types is not supported (see README):\n" +
-                s"Source Avro schema: $sourceAvroSchema.\n" +
-                s"Target Catalyst type: $targetSqlType")
+            case other =>
+              sqlType match {
+                case t: StructType if t.fields.length == avroSchema.getTypes.size =>
+                  val fieldConverters = t.fields zip avroSchema.getTypes map {
+                    case (field, schema) =>
+                      createConverter(schema, field.dataType, path :+ field.name)
+                  }
+
+                  (item: AnyRef) => if (item == null) {
+                    null
+                  } else {
+                    val i = GenericData.get().resolveUnion(avroSchema, item)
+                    val converted = new Array[Any](fieldConverters.length)
+                    converted(i) = fieldConverters(i)(item)
+                    new GenericRow(converted)
+                  }
+                case _ => throw new IncompatibleSchemaException(
+                  s"Cannot convert Avro schema to catalyst type because schema at path " +
+                    s"${path.mkString(".")} is not compatible " +
```

Comment:
Nit: I'd prefer "not supported" instead of "not compatible". "Compatible" is a term usually used in scenarios like schema evolution/merging, while here it means Spark SQL doesn't recognize some specific Avro schema.